[PVR] Confirm channel switch when flipping through channel informatio…

ksooo opened the pull request:
Confirm channel switch when flipping through channel information in fullscreen.

This feature has been requested several times in the forum, e.g. here: http://forum.kodi.tv/showthread.php?tid=196389

I've never before dealt with the code I changed for this PR, so, if my approach is bullshit, let me know. ;-)

As you can see a reused the exiting setting that currently activates confirmation of channel switch when using the numeric keys to select a channel when in fullscreen. Should there be a new setting for the new feature?

da-anda commented:
I don't think we need a second setting - if there are concerns the setting could be converted into a spinner with options like "off", "always", "on numeric input", "on channel up/down" (or alike).

ksooo commented:
@da-anda sounds good to me. I would extend the setting only in case people complain.

da-anda commented:
absolutely, let's wait on user feedback before complicating things

ksooo commented:
@xhaggi mind taking a look?

xhaggi commented:
I'll need some more time to review this.

ksooo commented:
@xhaggi np. Take your time. No need to rush. Just wanted to make sure you did not miss this.

xhaggi commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -5832,6 +5884,18 @@ bool CGUIInfoManager::ConditionsChangedValues(const std::map<INFO::InfoPtr, bool
+  return pOSD && pOSD->IsDialogRunning();
+}
+
+bool CGUIInfoManager::GetShowMusicOSD() const
```
you can use `CGUIWindowManager::IsWindowActive(int id)` instead of introducing these methods.

ksooo replied:
@xhaggi I'm calling this methods from outside <code>CGUIInfoManager</code>, for example here https://github.com/xbmc/xbmc/pull/7379/files#diff-14ed9651a2eb656ff75aea52d0912a00R4112 and wanted to encapsulate the implemantion. I consider <code>CDVDPlayer</code> dealing *directly* with gui window code no good. Thus, I would keep the methods but simplify the implemention. Okay?

xhaggi commented on xbmc/cores/dvdplayer/DVDPlayer.cpp:
```diff
@@ -4640,6 +4655,10 @@ std::string CDVDPlayer::GetPlayingTitle()
 bool CDVDPlayer::SwitchChannel(const CPVRChannelPtr &channel)
 {
+  const CPVRChannelPtr currentChannel(g_PVRManager.GetCurrentChannel());
+  if (channel && currentChannel && (*channel == *currentChannel))
```
it's enough to compare the shared_ptr's because there are operators for comparing a shared_ptr with a null pointer.

```
if (channel == currentChannel)
```

ksooo replied:
@xhaggi: AFAIK, <code>channel == currentchannel</code> would just compare the pointer values. I want to compare the actual channel objects.

xhaggi commented on xbmc/pvr/PVRManager.cpp:
```diff
@@ -1121,6 +1121,25 @@ bool CPVRManager::PlayMedia(const CFileItem& item)
+    CPVRChannelPtr channel(m_currentFile->GetPVRChannelInfoTag());
+    CPVRChannelPtr playingChannel(GetCurrentChannel());
+
+    if (channel && playingChannel && (*channel != *playingChannel))
```
same here

opdenkamp replied:
would probably make sense to just add a method `bool CPVRManager::IsPlayingChannel(const CPVRChannelPtr& channel)` because we have the same check in multiple places

opdenkamp commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -2145,6 +2147,36 @@ bool CGUIInfoManager::GetInt(int &value, int info, int contextWindow, const CGUI
+                time_t te;
+                end.GetAsTime(te);
+
+                value = static_cast<int>(((tn - ts) * 100)/(te - ts));
```
`CEpgInfoTag::ProgressPercentage()`?

opdenkamp commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -4396,9 +4432,25 @@ void CGUIInfoManager::UpdateFPS()
+    }
+  }
+  return bRet;
+}
```
bool CGUIInfoManager::PlayingChannelMatchesDisplayedChannel() const
{
  return g_application.m_pPlayer->IsPlaying() &&
         m_currentFile->HasPVRChannel() &&
         *m_currentFile->GetPVRChannelInfoTag() == *g_PVRManager.GetCurrentChannel(); // there will always be a channel in pvrmanager while m_currentFile has a channel
}

opdenkamp commented on xbmc/music/windows/GUIWindowVisualisation.cpp:
```diff
@@ -240,3 +241,32 @@ void CGUIWindowVisualisation::FrameMove()
+    {
+      pOSD->Close();
+    }
+    else
```
I'd rewrite this block too:
```C++
    else if (g_infoManager.GetShowInfo() &&
          CSettings::Get().GetBool("pvrplayback.confirmchannelswitch") &&
          g_application.CurrentFileItem().HasChannel())
    {
      // If only channel info but no player controls are currently visible
      // and setting "confirm channel switch with OK" is active, do a channel
      // switch instead of turning on the the player controls.
      g_application.m_pPlayer->SwitchChannel(g_application.CurrentFileItem().GetPVRChannelInfoTag());
      return;
    }
```

opdenkamp commented on xbmc/video/windows/GUIWindowFullScreen.cpp:
```diff
@@ -634,9 +635,26 @@ void CGUIWindowFullScreen::ToggleOSD()
+      // and setting "confirm channel switch with OK" is active, do a channel
+      // switch instead of turning on the the player controls.
+      if (g_infoManager.GetShowInfo() &&
+          CSettings::Get().GetBool("pvrplayback.confirmchannelswitch"))
```
similar to the other one above, can loose 2 unneeded scopes here

ksooo commented:
@opdenkamp thanks for the review. All valid points and suggestions. Will change code accordingly.

ksooo commented:
@xhaggi All feedback comments are now in. Some more or are we ready to (rebase and) merge?

ksooo commented:
@xhaggi Rethought things again. Came to the conclusion that my current approach to handle show/hide of certain OSD information directly in <code>CGUIInfoManager</code> using the newly introduced method <code>PlayingChannelMatchesDisplayedChannel()</code> does not really work.

For instance, "Currently recording" is shown in the OSD of the playing channel if that channel is currently recording. If OSD is in preview mode (aka not showing the currently playing channel), we obviously need to hide "Currently recording". Confluence uses "Player.Recording" to control whether to show or hide "Currently recording". In this PR in it's current state I changed implementation of "Player.Recording" to only return true <code>if (PlayingChannelMatchesDisplayedChannel())</code>. If let's say some other piece of kodi/skin/addon uses "Player.Recording" (and the OSD is in preview mode at that very moment) this would break that piece of code! No good, right? Same applies for other gui info implementations I changed. :-(

So, I think of another approach, to not change behavior of existing code in <code>CGUIInfoManager</code>, but (like it usually always works) let the skin decide what to show in the OSD and when. For this, I would add a new info to <code>CGUIInfoManager</code>, like "OSD.IsChannelPreview" that the skin can use to control visibility of items in preview mode. Drawback is that all skins must be adjusted, otherwise they would probably show to many items when in preview mode, but currently I see no other solution that actually works without changing skins.

What do you think?

ksooo commented:
@xhaggi could you please take a look at the impl of the new approach. Much cleaner code and actually not breaking other stuff (imo). Like it.

xhaggi commented:
sorry I'm very busy atm so I can't do that before Monday. would it make more sense to introduce a new pvr osd dialog which handle both radio and tv and the preview stuff instead hacky all in the current ones? much cleaner and separated from the rest.

ksooo commented:
@xhaggi np, take your time. 

I don't consider the new approach hacky, btw. It works like it always(?) does. There is a dialog and the skin controls what is shown and when - using guiinfo conditions. I thought about a new dialog, but came to the conclusion that the new one and the existing one would have very much in common, thus, skinners would have to take care to keep those dialogs in sync. No good, imo.  Error-prone.

ksooo commented:
@xhaggi would be great to get this to an end the next days as I will be not available for Kodi related todos from Aug 8 to Aug 30 and I really would like to get this feature in for Jarvis.

Plesae take into consideration that users have requested this feature for years(!). I use for quite some time now with my private Kodi build. It works like a charm and I never ever will be without this feature.

Finally, my feeling is that it would be good for the Jarvis feature list if the one or the other bullet point could be added. It's currently not that long, imo... ;-)

da-anda commented:
Hehe, you know that the feature list argument is a bit moot? :) I think the ADSP is already big enough of a new feature. Then add Retroplayer (ideally) and nobody will care about "confirm channel switch" in the feature list :P Never the less I'd also like to see it merged in upcoming window ;)

ksooo commented:
@da-anda this ofc all personal opinion. I have no idea how many users asked for ADSP or how far Retroplayer is away from being production ready. I only know the the confirm channel switch stuff was requested for a long time by a lot of users. And those users most probably will care, reagrdless what else is on the list.

xhaggi commented:
@ksooo will take a look at it tomorrow

ksooo commented:
@xhaggi great. Thanks in advance.

xhaggi commented on xbmc/music/windows/GUIWindowVisualisation.cpp:
```diff
@@ -240,3 +241,29 @@ void CGUIWindowVisualisation::FrameMove()
   CGUIWindow::FrameMove();
 }
+
+void CGUIWindowVisualisation::ToggleOSD()
```
could you please explain me what you want to achieve with that method?

ksooo replied:
Well, it also took me a while to understand this code (I hope I actually did). I'm afraid it will take some more words to explain things. :-/ So please take yourself the time to read through the following wall of text. ;-)

GUIWindowVisualisation and GUIWindowFullscreen to some extend are doing similar things, working almost the same, the first for music, the latter for video. Let me explain things on the music side:

GUIWindowVisualisation (among other things) is responsible for music related OSD, like title, start and end time and such, but not for player controls (explained later). The visibility of the information is managed by flipping a bool (GUIInfoManager::SetShowInfo) whenever a component of Kodi wants to show/hide the OSD. Confluence checks this bool ("Player.ShowInfo") in MusicVisualisation.xml and shows/hides stuff accordingly. 

GUIDialogMusicOSD implements and manages the OSD player controls, like start, stop, etc.

Thus, if you see an OSD with both information for the music and player controls, technically there is GUIWindowVisualisation having set Player.ShowInfo to true and a running DialogMusicOSD.

Putting the peaces together: GUIWindowVisualisation::ToggleOSD gets called to toggle the visibility state of the player controls only, not the visibility of the OSD music information. For this, it first looks whether the player controls are visible (=> DialogMusicOSD is running). If so, it closes the dialog to hide the player controls. If not it calls DialogMusicOSD::Open to show the player controls. 

The name ToggleOSD is misleading, if you ask me, as it actually manages only the visibility of the player controls. But naming was there for the video part and I reused/copied it for music. Maybe we should change the naming later in another PR.

The logic just described is not new with this PR. In the first place I just extended the music side to implement the toggle correctly, like the video side does (introduction of GUIWindowVisualisation::ToggleOSD, as a copy of GUIWindowFullscreen::ToggleOSD).

Then I added the logic for the confirm channel switch feature to GUIWindowVisualisation::ToggleOSD. Will explain this in another comment.

xhaggi commented on xbmc/music/windows/GUIWindowVisualisation.cpp:
```diff
@@ -240,3 +241,29 @@ void CGUIWindowVisualisation::FrameMove()
+  {
+    if (pOSD->IsDialogRunning())
+    {
+      pOSD->Close();
```
If you open the OSD with ACTION_SHOW_OSD this window lost the focus and the OSD will consume all messages. IMO this method is never called in case the the OSD is running.

ksooo replied:
Do not understand this, sorry. I can only say that it works and that this method gets called. ???

xhaggi commented on xbmc/music/windows/GUIWindowVisualisation.cpp:
```diff
@@ -240,3 +241,29 @@ void CGUIWindowVisualisation::FrameMove()
+      // If only channel info but no player controls are currently visible
+      // and setting "confirm channel switch with OK" is active, do a channel
+      // switch instead of turning on the the player controls.
+      g_application.m_pPlayer->SwitchChannel(g_application.CurrentFileItem().GetPVRChannelInfoTag());
```
not sure I understand what this is supposed to be.

ksooo replied:
This method gets called if the user hits OK while the OSD is displayed - the event we are interested in for confirm channel switch. If the player controls are not visible then and music information OSD is visible and confirm channel switch is activate and currently displayed informattion is for a PVR supplied radio channel (confirm channel switch only works for PVR TV/radio stuff), then and only then pressing OK switches to the channel currently displayed in OSD.

Reading this myself, I suddenly get the strong feeling that there is a better place for putting this logic. ;-) Any suggestions?

xhaggi replied:
I would say place it in `OnAction()` or call a method `GUIWindowVisualisation::SwitchChannel()` in `OnAction()` which returns true/false. The action should be `ACTION_SELECT_ITEM` because `ACTION_SHOW_GUI` is mapped to "fullscreen" in `Key.h

ksooo replied:
@xhaggi: Does not work. And now I remember... Had it in the first place almost like you suggested and then found out that when hitting OK on my remote no ACTION_SELECT_ITEM, but instead ACTION_SHOW_OSD was created. Reason is here:
https://github.com/xbmc/xbmc/blob/master/system/keymaps/remote.xml#L196 and here https://github.com/xbmc/xbmc/blob/master/system/keymaps/remote.xml#L242
When changing the lines in question to <code>\<select\>Select\</select\></code> it works just fine with ACTION_SELECT_ITEM. So, the question is whether it is okay to change remote.xml that way? I'd say yes.

xhaggi replied:
mhh my remote.xml holds `<select>Playlist</select>` which open the channel osd. The issue here is that the info control (info overlay) depends on `Player.ShowInfo` and is no real dialog where we could catch the right action.

ksooo replied:
This is why I think that my approach to hook the conditional channel switch into ACTION_SHOW_OSD is not that much wrong. What do you say on second thought?

xhaggi replied:
no it won't work, because with my remote.xml you will end-up with the same issue. I can't confirm the channel switch with "OK" on my remote, because it sends `ACTION_SHOW_PLAYLIST`.

xhaggi replied:
As I said before, best we introduce a new dialog which becomes active if you switch a channel (tv/radio) and display the related channel info. Within this dialog you can handle the action `ACTION_SELECT_ITEM` and do the necessary stuff. There is no way to handle what you are currently want to achieve within the FullscreenVideo/Visualisation window because the OK/Enter key is mapped differently depends on the use-case. Keep in mind that you need to handle channel up/down in this new dialog too. There is no fallback to fullscreen/visualisation window's key mapping. Another advantage you can cancel/close this dialog (hide the channel info) with back, which is also not possible in the current implementation. You need to wait until the info control disappears.

da-anda commented:
@ksooo I hope you didn't take my comment about the feature list too serious

ksooo commented:
@da-anda nvm. "Everything is awesome!!!" www.youtube.com/watch?v=StTqXEQ2l-Y

ksooo commented:
@xhaggi took a deeper look into the approach with the new dialog you suggested. Turned out to me that this is really very much work as the channel info control and switch logics today is spread over several source files (dvdplayer, dvdplayer, dvdinputsreammanager, pvrmamanager, pvrguiinfo, ...). Implementing the new dialog is really huge effort, imo. This would end up in a bigger refactoring before we could start to implement this feature, imo.

OTOH it is quite easy to solve the "select has not the semantics of OK in FullscreenVideo and/or Visualisation window in my configuration" you mentioned. My latest commit contains a working pragmatic solution with not to much code changes (regression risk reduced), for your consideration.

I know that this, from the academic point of view, is not an ideal solution but it works and would get Kodi users finally the feature they are waiting for a long time now.

xhaggi commented on xbmc/input/ButtonTranslator.h:
```diff
@@ -78,6 +78,7 @@ class CButtonTranslator
   static void GetActions(std::vector<std::string> &actionList);
   static void GetWindows(std::vector<std::string> &windowList);
 
+  CAction GetGlobalAction(const CKey &key) const;
```
@ksooo please add a method comment.

xhaggi commented on xbmc/input/ButtonTranslator.cpp:
```diff
@@ -1170,6 +1170,12 @@ int CButtonTranslator::GetFallbackWindow(int windowID)
+CAction CButtonTranslator::GetGlobalAction(const CKey &key) const
+{
+  std::string strAction;
+  return GetActionCode(-1, key, strAction);
```
@ksooo please use `CButtonTranslator::GetActionCode(int window, int action)` to get rid of the dummy string var.

xhaggi replied:
meh sorry my mistake. i meant `CButtonTranslator::GetAction(int window, const CKey &key, bool fallback)`.

ksooo replied:
You certainly mean this (https://github.com/xbmc/xbmc/blob/master/xbmc/input/ButtonTranslator.h#L116), right?

ksooo committed:
[PVR] Confirm channel switch when flipping through channel information in fullscreen.

ksooo commented:
@xhaggi all review comments are now in. Gimme your +1, please.

jenkins build this please

xhaggi commented on xbmc/cores/dvdplayer/DVDPlayer.cpp:
```diff
@@ -2593,7 +2593,8 @@ void CDVDPlayer::HandleMessages()
           bool bSwitchSuccessful(false);
-          bool bShowPreview(CSettings::Get().GetInt("pvrplayback.channelentrytimeout") > 0);
+          bool bShowPreview(!g_infoManager.IsPlayerOSDActive() &&
+                            (CSettings::Get().GetInt("pvrplayback.channelentrytimeout") > 0));
```
this won't work if you've not set the channel entry timeout but the confirm channel switch option.

This should be:
```
bool bShowPreview(!g_infoManager.IsPlayerOSDActive() &&
                            (CSettings::Get().GetBool("pvrplayback.confirmchannelswitch") ||
                             CSettings::Get().GetInt("pvrplayback.channelentrytimeout") > 0));
```

xhaggi commented:
please adjust the condition and it should be fine :smile:

ksooo commented:
Condition fixed.

wsnipex commented:
jenkins build this please

ksooo commented:
jenkins build this please

xhaggi commented:
jenkins build and merge

xhaggi commented:
build error is unrelated
